smb: add support for SMB

Lesmiscore opened the pull request:
<!--
Thank you very much for contributing code or documentation to rclone! Please
fill out the following questions to make it easier for us to review your
changes.

You do not need to check all the boxes below all at once, feel free to take
your time and add more commits. If you're done and ready for review, please
check the last box.
-->

#### What is the purpose of this change?

Add the SMB backend

#### Was the change discussed in an issue or in the forum before?

Closes #2042

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

albertony commented:
Interesting! 👍 

Maybe we should cross-link its doc page with the [local backend](https://rclone.org/local/), which today mentions:

> Network paths in UNC format, \\\\server\\share, are also supported.

I assume this SMB backend will overlap with that? And then what are the (dis-)advantages with either of them?

Lesmiscore commented:
> Maybe we should cross-link its doc page with the [local backend](https://rclone.org/local/), which today mentions:
> 
> > Network paths in UNC format, \\server\share, are also supported.

I guess so



> I assume this SMB backend will overlap with that? And then what are the (dis-)advantages with either of them?

The benefit for this is that it works without Windows or GVFS installation, even on embedded devices or CIs

albertony commented on backend/smb/smb.go:
```diff
@@ -0,0 +1,790 @@
+			Advanced: true,
+		}, {
+			Name:     "case_insensitive",
+			Help:     "Whether the server is configured to be case-insensitive. Always true on Windows shares.",
```
I think this one will be better if either writing as a single sentence, e.g. `,` instead of `.` before `Always true...`, or if splitting into two paragraphs, i.e. inserting `\n\n`. This has to do with different uses of such help texts, and especially `rclone help flags` which extracts first line and adds information about default value to it. Your other help texts looks fine.

Lesmiscore replied:
I'll go with splitting it into two paragraphs

ncw commented:
This is a great piece of work - thank you :-)

How are the integration tests looking?

Do you think you could make a docker image so we can run integration tests automatically, something like

https://github.com/rclone/rclone/blob/master/fstest/testserver/init.d/TestFTPProftpd

?

Maybe using something like this? https://hub.docker.com/r/dperson/samba

Lesmiscore commented:
> How are the integration tests looking?

```
$ go test -v backend/smb/smb_test.go -args --verbose --remote samba-test:extrastorage/sambatransfer/rcloneintegrationtest/
=== RUN   TestIntegration
    fstests.go:454: Using remote "samba-test:extrastorage/sambatransfer/rcloneintegrationtest/"
2022/09/20 13:22:35 DEBUG : Creating backend with remote "samba-test:extrastorage/sambatransfer/rcloneintegrationtest//rclone-test-kuyujof0wumanig8fiqapal6"
2022/09/20 13:22:35 DEBUG : Using config file from "/home/lesmi/.config/rclone/rclone.conf"
=== RUN   TestIntegration/FsCheckWrap
    fstests.go:495: Not a wrapping Fs
=== RUN   TestIntegration/FsCommand
    fstests.go:523: No commands in this remote
=== RUN   TestIntegration/FsRmdirNotFound
    fstests.go:538: Skipping test as non root bucket-based remote
=== RUN   TestIntegration/FsString
=== RUN   TestIntegration/FsName
=== RUN   TestIntegration/FsRoot
=== RUN   TestIntegration/FsRmdirEmpty
=== RUN   TestIntegration/FsMkdir
=== RUN   TestIntegration/FsMkdir/FsMkdirRmdirSubdir
2022/09/20 13:22:35 DEBUG : dir/subdir: Making directory
2022/09/20 13:22:35 INFO  : dir/subdir: Removing directory
2022/09/20 13:22:35 INFO  : dir: Removing directory
=== RUN   TestIntegration/FsMkdir/FsListEmpty
=== RUN   TestIntegration/FsMkdir/FsListDirEmpty
=== RUN   TestIntegration/FsMkdir/FsListRDirEmpty
    fstests.go:415: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsListDirNotFound
2022/09/20 13:22:35 ERROR : does not exist: error listing: directory not found
=== RUN   TestIntegration/FsMkdir/FsListRDirNotFound
    fstests.go:415: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsEncoding
=== RUN   TestIntegration/FsMkdir/FsEncoding/control_chars
    fstests.go:710: testing "␀␁␂␃␄␅␆␇␈␉␊␋␌␍␎␏␐␑␒␓␔␕␖␗␘␙␚␛␜␝␞␟␡"
=== RUN   TestIntegration/FsMkdir/FsEncoding/dot
    fstests.go:710: testing "．"
=== RUN   TestIntegration/FsMkdir/FsEncoding/dot_dot
    fstests.go:710: testing "．．"
=== RUN   TestIntegration/FsMkdir/FsEncoding/punctuation
    fstests.go:710: testing "!\"#$%&'()*+,-.／:;<=>?@[\\]^_`{|}~"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_space
    fstests.go:710: testing " leading space"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_tilde
    fstests.go:710: testing "~leading tilde"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_CR
    fstests.go:710: testing "␍leading CR"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_LF
    fstests.go:710: testing "␊leading LF"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_HT
    fstests.go:710: testing "␉leading HT"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_VT
    fstests.go:710: testing "␋leading VT"
=== RUN   TestIntegration/FsMkdir/FsEncoding/leading_dot
    fstests.go:710: testing ".leading dot"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_space
    fstests.go:710: testing "trailing space "
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_CR
    fstests.go:710: testing "trailing CR␍"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_LF
    fstests.go:710: testing "trailing LF␊"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_HT
    fstests.go:710: testing "trailing HT␉"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_VT
    fstests.go:710: testing "trailing VT␋"
=== RUN   TestIntegration/FsMkdir/FsEncoding/trailing_dot
    fstests.go:710: testing "trailing dot."
=== RUN   TestIntegration/FsMkdir/FsEncoding/invalid_UTF-8
    fstests.go:710: testing "invalid utf-8\xfe"
=== RUN   TestIntegration/FsMkdir/FsEncoding/URL_encoding
    fstests.go:710: testing "test%46.txt"
=== RUN   TestIntegration/FsMkdir/FsNewObjectNotFound
=== RUN   TestIntegration/FsMkdir/FsPutError
2022/09/20 13:22:36 DEBUG : hello? sausage/êé/Hello, 世界/ " ' @ < > & ? + ≠/z.txt: removed after failed upload: potato
=== RUN   TestIntegration/FsMkdir/FsPutZeroLength
=== RUN   TestIntegration/FsMkdir/FsOpenWriterAt
    fstests.go:783: FS has no OpenWriterAt interface
=== RUN   TestIntegration/FsMkdir/FsChangeNotify
    fstests.go:819: FS has no ChangeNotify interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListDirFile2
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRDirFile2
    fstests.go:415: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListR
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRSubdir
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListDirRoot
2022/09/20 13:22:36 DEBUG : Creating backend with remote "samba-test:extrastorage/sambatransfer/rcloneintegrationtest/"
2022/09/20 13:22:37 DEBUG : smb://guest@100.123.26.48:445/extrastorage/sambatransfer/rcloneintegrationtest: Using root directory "extrastorage/sambatransfer/rcloneintegrationtest"
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRDirRoot
    fstests.go:415: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListSubdir
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRSubdir#01
    fstests.go:415: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListLevel2
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListRLevel2
    fstests.go:415: FS has no ListR interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListFile1
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsNewObject
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsNewObjectCaseInsensitive
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsNewObjectCaseInsensitive/Dir
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsListFile1and2
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsNewObjectDir
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsPurge
    fstests.go:1104: FS has no Purge interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsCopy
    fstests.go:1141: FS has no Copier interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsMove
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsDirMove
2022/09/20 13:22:37 DEBUG : Creating backend with remote "samba-test:extrastorage/sambatransfer/rcloneintegrationtest//rclone-test-gogoruc6wokihub8fizutik5"
2022/09/20 13:22:37 DEBUG : smb://guest@100.123.26.48:445/extrastorage/sambatransfer/rcloneintegrationtest/rclone-test-gogoruc6wokihub8fizutik5: Purge dir "new_name"
2022/09/20 13:22:37 DEBUG : smb://guest@100.123.26.48:445/extrastorage/sambatransfer/rcloneintegrationtest/rclone-test-gogoruc6wokihub8fizutik5: Purge dir ""
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsRmdirFull
    fstests.go:1301: Skipping test as non root bucket-based remote
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsPrecision
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectString
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectFs
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectRemote
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectHashes
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectModTime
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectMimeType
    fstests.go:1379: MimeType method not supported
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectMetadata
    fstests.go:1408: Metadata method not supported
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectSetModTime
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectSize
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectOpen
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectOpenSeek
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectPartialRead
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectUpdate
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectStorable
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsIsFile
2022/09/20 13:22:37 DEBUG : Creating backend with remote "samba-test:extrastorage/sambatransfer/rcloneintegrationtest//rclone-test-kuyujof0wumanig8fiqapal6/hello? sausage/êé/Hello, 世界/ \" ' @ < > & ? + ≠/z.txt"
2022/09/20 13:22:37 DEBUG : smb://guest@100.123.26.48:445/extrastorage/sambatransfer/rcloneintegrationtest/rclone-test-kuyujof0wumanig8fiqapal6/hello? sausage/êé/Hello, 世界/ " ' @ < > & ? + ≠: Using root directory "extrastorage/sambatransfer/rcloneintegrationtest/rclone-test-kuyujof0wumanig8fiqapal6/hello? sausage/êé/Hello, 世界/ \" ' @ < > & ? + ≠"
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsIsFileNotFound
2022/09/20 13:22:37 DEBUG : Creating backend with remote "samba-test:extrastorage/sambatransfer/rcloneintegrationtest//rclone-test-kuyujof0wumanig8fiqapal6/not found.txt"
2022/09/20 13:22:37 ERROR : : error listing: directory not found
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FromRoot
    fstests.go:1601: Can't list from root on this remote
=== RUN   TestIntegration/FsMkdir/FsPutFiles/PublicLink
    fstests.go:1705: FS has no PublicLinker interface
=== RUN   TestIntegration/FsMkdir/FsPutFiles/SetTier
    fstests.go:428: FS has no SetTier & GetTier interfaces
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectCheckWrap
    fstests.go:1808: Not a wrapping Fs
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectRemove
=== RUN   TestIntegration/FsMkdir/FsPutFiles/ObjectAbout
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsPutStream
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsPutStream/0
=== RUN   TestIntegration/FsMkdir/FsPutFiles/FsPutStream/100
=== RUN   TestIntegration/FsMkdir/FsPutFiles/Internal
    fstests.go:1904: *smb.Fs does not implement InternalTester
=== RUN   TestIntegration/FsMkdir/FsPutChunked
    fstests.go:1921: *smb.Fs does not implement SetUploadChunkSizer
=== RUN   TestIntegration/FsMkdir/FsUploadUnknownSize
=== RUN   TestIntegration/FsMkdir/FsUploadUnknownSize/FsPutUnknownSize
=== RUN   TestIntegration/FsMkdir/FsUploadUnknownSize/FsUpdateUnknownSize
=== RUN   TestIntegration/FsMkdir/FsRootCollapse
2022/09/20 13:22:37 DEBUG : Creating backend with remote "samba-test:extrastorage/sambatransfer/rcloneintegrationtest//rclone-test-kuyujof0wumanig8fiqapal6/deeper/nonexisting/directory"
2022/09/20 13:22:37 DEBUG : Waiting for deletions to finish
2022/09/20 13:22:37 INFO  : hello? sausage/êé/Hello, 世界/ " ' @ < > & ? + ≠/z.txt: Deleted
2022/09/20 13:22:37 INFO  : hello? sausage/êé/Hello, 世界/ " ' @ < > & ? + ≠: Removing directory
2022/09/20 13:22:37 INFO  : hello? sausage/êé/Hello, 世界: Removing directory
2022/09/20 13:22:37 INFO  : hello? sausage/êé: Removing directory
2022/09/20 13:22:37 INFO  : hello? sausage: Removing directory
2022/09/20 13:22:37 INFO  : smb://guest@100.123.26.48:445/extrastorage/sambatransfer/rcloneintegrationtest/rclone-test-kuyujof0wumanig8fiqapal6: Removing directory
2022/09/20 13:22:37 ERROR : : error listing: directory not found
=== RUN   TestIntegration/FsShutdown
2022/09/20 13:22:37 DEBUG : smb://guest@100.123.26.48:445/extrastorage/sambatransfer/rcloneintegrationtest/rclone-test-kuyujof0wumanig8fiqapal6: Closing 2 unused connections
--- PASS: TestIntegration (1.85s)
    --- SKIP: TestIntegration/FsCheckWrap (0.00s)
    --- SKIP: TestIntegration/FsCommand (0.00s)
    --- SKIP: TestIntegration/FsRmdirNotFound (0.00s)
    --- PASS: TestIntegration/FsString (0.00s)
    --- PASS: TestIntegration/FsName (0.00s)
    --- PASS: TestIntegration/FsRoot (0.00s)
    --- PASS: TestIntegration/FsRmdirEmpty (0.01s)
    --- PASS: TestIntegration/FsMkdir (1.74s)
        --- PASS: TestIntegration/FsMkdir/FsMkdirRmdirSubdir (0.11s)
        --- PASS: TestIntegration/FsMkdir/FsListEmpty (0.01s)
        --- PASS: TestIntegration/FsMkdir/FsListDirEmpty (0.01s)
        --- SKIP: TestIntegration/FsMkdir/FsListRDirEmpty (0.00s)
        --- PASS: TestIntegration/FsMkdir/FsListDirNotFound (0.01s)
        --- SKIP: TestIntegration/FsMkdir/FsListRDirNotFound (0.00s)
        --- PASS: TestIntegration/FsMkdir/FsEncoding (0.81s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/control_chars (0.12s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/dot (0.04s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/dot_dot (0.03s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/punctuation (0.03s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_space (0.02s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_tilde (0.05s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_CR (0.06s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_LF (0.06s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_HT (0.03s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_VT (0.02s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/leading_dot (0.02s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_space (0.02s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_CR (0.02s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_LF (0.03s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_HT (0.05s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_VT (0.05s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/trailing_dot (0.06s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/invalid_UTF-8 (0.05s)
            --- PASS: TestIntegration/FsMkdir/FsEncoding/URL_encoding (0.03s)
        --- PASS: TestIntegration/FsMkdir/FsNewObjectNotFound (0.00s)
        --- PASS: TestIntegration/FsMkdir/FsPutError (0.01s)
        --- PASS: TestIntegration/FsMkdir/FsPutZeroLength (0.01s)
        --- SKIP: TestIntegration/FsMkdir/FsOpenWriterAt (0.00s)
        --- SKIP: TestIntegration/FsMkdir/FsChangeNotify (0.00s)
        --- PASS: TestIntegration/FsMkdir/FsPutFiles (0.65s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListDirFile2 (0.03s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsListRDirFile2 (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListR (0.04s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListRSubdir (0.03s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListDirRoot (0.04s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsListRDirRoot (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListSubdir (0.02s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsListRSubdir#01 (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListLevel2 (0.01s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsListRLevel2 (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListFile1 (0.04s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsNewObject (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsNewObjectCaseInsensitive (0.01s)
                --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsNewObjectCaseInsensitive/Dir (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsListFile1and2 (0.02s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsNewObjectDir (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsPurge (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsCopy (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsMove (0.09s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsDirMove (0.07s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FsRmdirFull (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsPrecision (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectString (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectFs (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectRemote (0.02s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectHashes (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectModTime (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/ObjectMimeType (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/ObjectMetadata (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectSetModTime (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectSize (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectOpen (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectOpenSeek (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectOpenRange (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectPartialRead (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectUpdate (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectStorable (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsIsFile (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsIsFileNotFound (0.01s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/FromRoot (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/PublicLink (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/SetTier (0.00s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/ObjectCheckWrap (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectRemove (0.02s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/ObjectAbout (0.00s)
            --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsPutStream (0.02s)
                --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsPutStream/0 (0.01s)
                --- PASS: TestIntegration/FsMkdir/FsPutFiles/FsPutStream/100 (0.01s)
            --- SKIP: TestIntegration/FsMkdir/FsPutFiles/Internal (0.00s)
        --- SKIP: TestIntegration/FsMkdir/FsPutChunked (0.00s)
        --- PASS: TestIntegration/FsMkdir/FsUploadUnknownSize (0.03s)
            --- PASS: TestIntegration/FsMkdir/FsUploadUnknownSize/FsPutUnknownSize (0.01s)
            --- PASS: TestIntegration/FsMkdir/FsUploadUnknownSize/FsUpdateUnknownSize (0.02s)
        --- PASS: TestIntegration/FsMkdir/FsRootCollapse (0.01s)
    --- PASS: TestIntegration/FsShutdown (0.00s)
PASS
ok      command-line-arguments  1.857s
```
Everything is okay



> Do you think you could make a docker image so we can run integration tests automatically, something like

> Maybe using something like this? [hub.docker.com/r/dperson/samba](https://hub.docker.com/r/dperson/samba)

I think that docker image works. The tests above are carried out by using the servers with smbd

ncw commented:
Nice!

Do you want to add the docker image to `fstest/testserver/init.d/TestSMB` and add it to fstest/test_all/config.yml ?

I can do that if you want.

Other than that, do you think the code is ready to merge?

Lesmiscore commented:
> Do you want to add the docker image to `fstest/testserver/init.d/TestSMB` and add it to fstest/test_all/config.yml ?

Can you please make one for it? I'm not familiar with how these tests are done


> Other than that, do you think the code is ready to merge?

Yes

ncw commented:
I've added a commit which fixes these

- `password` should be called `pass` to be consistent with other backends and needs to have `obscure.Reveal` run on it
- `username` should be called `user` to be consistent with other backends 

I've added another commit which adds a docker command which runs a local smb server in a docker container. This will run if you do `go test -v` or `test_all -backends smb`

I added a 3rd commit to fix the lint problem.

All the tests pass - super job - well done :-)

I'll give the code a review now.

ncw commented on backend/smb/smb.go:
```diff
@@ -0,0 +1,790 @@
+}
+
+type boundReadCloser struct {
+	rc    io.Reader
```
This should probably be `rc io.ReadCloser`

ncw commented on backend/smb/smb.go:
```diff
@@ -0,0 +1,790 @@
+	return r.rc.Read(p)
+}
+
+func (r *boundReadCloser) Close() error {
```
I would just return the first error here - I don't really like the multi-error thing, many errors confuse the user!

Assuming you've turned rc into a `ReadCloser` as above then I think this should be

```go
err1 := r.rc.Close()
err2 := r.close()
if err1 != nil {
    return err1
}
return err2
```

Which stops this backend depending on the union backend which is good.

ncw commented on docs/content/smb.md:
```diff
@@ -0,0 +1,231 @@
+
+You can't access to the shared printers from rclone, obviously.
+
+You can't use Anonymous access for logging in. You have to use the `guest` user with an empty password instead. (Setting empty password requires tweaking `rclone.conf` file)
```
You should be able to set an empty password like this

```
Option pass.
SMB password.
Choose an alternative below. Press Enter for the default (n).
y) Yes, type in my own password
g) Generate random password
n) No, leave this optional password blank (default)
y/g/n> n
```

Which leaves out the password field which means it defaults to ""

ncw commented on docs/content/smb.md:
```diff
@@ -0,0 +1,231 @@
+You can't access to the shared printers from rclone, obviously.
+
+You can't use Anonymous access for logging in. You have to use the `guest` user with an empty password instead. (Setting empty password requires tweaking `rclone.conf` file)
+The rclone client tries to avoid 8.3 names when uploading files. If you wish to prevent it, it is highly recommended to use either [FTP](/ftp) or [SFTP](/sftp) instead.
```
Have we done enough to prevent 8.3 names? How might they happen?

It would be nice to remove "If you wish to prevent it, it is highly recommended to use either [FTP](/ftp) or [SFTP](/sftp) instead." if we can

Lesmiscore replied:
The situation happens when the file name ends with spaces or dots (e.g. `rclone ` or `smb.go.`)
As SMB itself doesn't seem to have a way to convert it back to non-8.3 name, the encoding parameter has `EncodeRightSpace` and `EncodeRightPeriod`
It is the same for Windows - you'll see 8.3 names for these files too

ncw replied:
Ah, so we have fixed the problem with the encoding? I noticed this problem when I ran the tests without the encodings by accident.

Perhaps we should change the docs to

> The rclone client tries to avoid 8.3 names when uploading files by encoding trailing spaces and periods.

And drop the 

> If you wish to prevent it, it is highly recommended to use either [FTP](/ftp) or [SFTP](/sftp) instead.

Which I think will confuse users.

ncw reviewed (commented):
Code looks great :+1: 

I put some minor comments inline.

Lesmiscore commented on backend/smb/smb.go:
```diff
@@ -0,0 +1,790 @@
+
+/// Misc
+
+// split returns bucket and bucketPath from the rootRelativePath
```
yes

Lesmiscore commented on backend/smb/connpool.go:
```diff
@@ -0,0 +1,228 @@
+		return nil
+	}
+	if c.smbShare != nil {
+		err = c.smbShare.Umount()
```
```suggestion
		err = c.smbShare.Umount()
		c.smbShare = nil
```

Lesmiscore commented:
@ncw ping

Lesmiscore commented:
@ncw Ping again

ncw reviewed (approved):
I think this is looking really good - all the tests pass.

The docs could do with filling out a bit (@albertony maybe you could help here?) but we can do that after merge.

Thank you for a great feature
